hicommonwealth/commonwealth
# Migration topic thresholds to group IDs + Implement Groups controller

## Link to Issue
Closes: #5027
Closes: #5028
Closes: #5029

## Description of Changes
- The PR is entirely additive and does NOT mutate any existing routes, columns or logic.
- Adds a migration which adds a `group_ids` column to Topics, and creates new Groups based on existing topic thresholds.
- Implements CRUD + refresh logic in Groups controller
- Adds Zod library for request validation

## Test Plan
- Add local env `GATING_API_ENABLED=true`
- Run the migration locally, confirm that it created groups based on existing topic thresholds
- Make request `GET http://localhost:8080/api/groups?include_members=true&chain_id=ankr-forum`
  - confirm that it returns a list of groups
- Make request `PUT http://localhost:8080/api/refresh-membership`
  - body `{author_chain, chain, address, jwt}`
  - confirm that results show the allowed/rejected status for each topic
- Make request `DELETE http://localhost:8080/api/groups/XXX` where XXX is the group ID
  - body `{author_chain, chain, address, jwt}`
  - confirm that the group is deleted (data response data)
- Make request `POST http://localhost:8080/api/groups`
  - body `{author_chain, chain, address, jwt, metadata, requirements}`
  - confirm that a new group is created
- Make request `PUT http://localhost:8080/api/groups/XXX` where XXX is the group ID
  - body `{author_chain, chain, address, jwt, metadata, requirements}`
  - confirm that the group is updated

## Deployment Plan
- Add production env `GATING_API_ENABLED=true`
- Must run migration which generate new groups from topic thresholds

## Other Considerations
- This PR does NOT include any member group validation (via TBC), but it does wire up the function, which just resolves to "yes the user meets the requirement"
